src/saturn_engine/worker/job.py

```diff
@@ -25,38 +38,49 @@ def __init__(
         self.logger = getLogger(__name__, self)
         self.inventory = inventory
         self.services = services
-        self.batch_size = 10
         self.queue_item = queue_item
         self.state_service = services.cast_service(JobStateService)
+        self.config = LazyConfig([self.services.s.config.r, self.queue_item.config])
 
     async def run(self) -> AsyncGenerator[TopicOutput, None]:
         try:
-            cursor = self.queue_item.state.cursor
-            done = None
-
-            async with alib.scoped_iter(
-                self.inventory.iterate(after=cursor)
-            ) as iterator:
-                while done is not True:
-                    try:
-                        done = True
-                        async for item in alib.islice(iterator, self.batch_size):
-                            cursor = item.cursor
-                            done = False
-                            message = TopicMessage(
-                                id=MessageId(item.id),
-                                args=item.args,
-                                tags=item.tags,
-                                metadata=item.metadata,
-                            )
-                            yield message
-                    finally:
-                        if not done and cursor is not None:
-                            self.state_service.set_job_cursor(
-                                self.queue_item.name, cursor=cursor
-                            )
+            async with alib.scoped_iter(self._make_iterator()) as iterator:
+                async for item in iterator:
+                    cursor = item.cursor
+                    message = TopicMessage(
+                        id=MessageId(item.id),
+                        args=item.args,
+                        tags=item.tags,
+                        metadata=item.metadata,
+                    )
+
+                    yield message
+
+                    if cursor:
+                        self.state_service.set_job_cursor(
+                            self.queue_item.name, cursor=cursor
+                        )
 
             self.state_service.set_job_completed(self.queue_item.name)
         except Exception as e:
             self.logger.exception("Exception raised from job")
             self.state_service.set_job_failed(self.queue_item.name, error=e)
+
+    def _make_iterator(self) -> t.AsyncIterator[Item]:
+        cursor = self.queue_item.state.cursor
+        iterator = self.inventory.iterate(after=cursor)
+
+        # If we enable cursors states, we are going to decorate the iterator
+        # to buffer multiple messages together and load their states.
+        if self.options.enable_cursors_states:
+            buffered_iterator = iterators.async_buffered(
+                iterator,
+                flush_after=self.options.buffer_flush_after,
+                buffer_size=self.options.buffer_size,
+            )
```

Comment:
All that is left is to add a "middleware" here that will load the cursors states (Eventually I would like to refactor the Topics to have middleware builtin, but enough refactoring for now.